File: Unity/Scripts/Managers/KingdomManager.cs
using UnityEngine;
using System.Collections.Generic;
using System.Linq;

/// <summary>
/// KingdomManager - Manages all 8 kingdoms and their restoration
/// </summary>
public class KingdomManager : MonoBehaviour
{
    [Header("Kingdoms")]
    public List<Kingdom> kingdoms = new List<Kingdom>();
    
    void Start()
    {
        InitializeKingdoms();
        Debug.Log($"🏰 Kingdom Manager initialized with {kingdoms.Count} kingdoms");
    }
    
    void InitializeKingdoms()
    {
        kingdoms.Clear();
        
        // 1. Sylvara - Nature Kingdom
        kingdoms.Add(new Kingdom
        {
            id = "sylvara",
            name = "Sylvara",
            element = KingdomElement.NATURE,
            status = KingdomStatus.FALLEN,
            restorationProgress = 0f,
            lore = "A thriving woodland empire corrupted by Shadow cultists"
        });
        
        // 2. Pyrrathia - Fire Kingdom
        kingdoms.Add(new Kingdom
        {
            id = "pyrrathia",
            name = "Pyrrathia",
            element = KingdomElement.FIRE,
            status = KingdomStatus.FALLEN,
            restorationProgress = 0f,
            lore = "A volcanic empire destroyed by catastrophic Water flood"
        });
        
        // 3. Frostvale - Ice Kingdom
        kingdoms.Add(new Kingdom
        {
            id = "frostvale",
            name = "Frostvale",
            element = KingdomElement.ICE,
            status = KingdomStatus.FALLEN,
            restorationProgress = 0f,
            lore = "A crystalline ice kingdom frozen by Eternal Winter spell"
        });
        
        // 4. Zerath Dunes - Desert Kingdom
        kingdoms.Add(new Kingdom
        {
            id = "zerath_dunes",
            name = "Zerath Dunes",
            element = KingdomElement.DESERT,
            status = KingdomStatus.FALLEN,
            restorationProgress = 0f,
            lore = "A golden desert empire buried by massive sandstorms"
        });
        
        // 5. Mor'gahl Fen - Poison Kingdom
        kingdoms.Add(new Kingdom
        {
            id = "morgahl_fen",
            name = "Mor'gahl Fen",
            element = KingdomElement.POISON,
            status = KingdomStatus.FALLEN,
            restorationProgress = 0f,
            lore = "A mysterious swamp civilization destroyed by poison disaster"
        });
        
        // 6. Luminaris - Light Kingdom
        kingdoms.Add(new Kingdom
        {
            id = "luminaris",
            name = "Luminaris",
            element = KingdomElement.LIGHT,
            status = KingdomStatus.FALLEN,
            restorationProgress = 0f,
            lore = "A radiant crystal kingdom shattered by light explosion"
        });
        
        // 7. Umbral Reach - Shadow Kingdom
        kingdoms.Add(new Kingdom
        {
            id = "umbral_reach",
            name = "Umbral Reach",
            element = KingdomElement.SHADOW,
            status = KingdomStatus.FALLEN,
            restorationProgress = 0f,
            lore = "A dark kingdom pulled into void dimension"
        });
        
        // 8. Crown Convergence - Final Kingdom
        kingdoms.Add(new Kingdom
        {
            id = "crown_convergence",
            name = "Crown Convergence",
            element = KingdomElement.UNIFIED,
            status = KingdomStatus.FALLEN,
            restorationProgress = 0f,
            lore = "The original kingdom where the Crown was forged"
        });
    }
    
    public Kingdom GetKingdom(string kingdomId)
    {
        return kingdoms.Find(k => k.id == kingdomId);
    }
    
    public void DiscoverKingdom(string kingdomId)
    {
        Kingdom kingdom = GetKingdom(kingdomId);
        
        if (kingdom == null)
        {
            Debug.LogError($"❌ Kingdom not found: {kingdomId}");
            return;
        }
        
        if (kingdom.status == KingdomStatus.FALLEN)
        {
            kingdom.status = KingdomStatus.DISCOVERED;
            Debug.Log($"🏰 Discovered {kingdom.name}!");
            Debug.Log($"   {kingdom.lore}");
            
            // Show notification
            if (GameManager.Instance?.uiManager != null)
            {
                GameManager.Instance.uiManager.ShowKingdomNotification(kingdom.name, "Kingdom Discovered!");
            }
        }
    }
    
    public void ClearVault(string kingdomId)
    {
        Kingdom kingdom = GetKingdom(kingdomId);
        
        if (kingdom == null) return;
        
        if (!kingdom.vaultCleared)
        {
            kingdom.vaultCleared = true;
            kingdom.status = KingdomStatus.VAULT_CLEARED;
            kingdom.restorationProgress = 33f;
            
            Debug.Log($"👑 {kingdom.name} vault cleared!");
            Debug.Log($"   Restoration: {kingdom.restorationProgress}%");
            
            // Give rewards
            if (GameManager.Instance?.player != null)
            {
                GameManager.Instance.player.GainXP(5000);
                GameManager.Instance.player.AddGold(2000);
                GameManager.Instance.player.AddCrownShards(1);
            }
            
            // Show notification
            if (GameManager.Instance?.uiManager != null)
            {
                GameManager.Instance.uiManager.ShowKingdomNotification(kingdom.name, "Vault Cleared!");
            }
        }
    }
    
    public void RestoreKingdom(string kingdomId, float amount)
    {
        Kingdom kingdom = GetKingdom(kingdomId);
        
        if (kingdom == null) return;
        
        kingdom.restorationProgress = Mathf.Min(100f, kingdom.restorationProgress + amount);
        
        if (kingdom.restorationProgress >= 100f && kingdom.status != KingdomStatus.RESTORED)
        {
            kingdom.status = KingdomStatus.RESTORED;
            Debug.Log($"✨ {kingdom.name} fully restored!");
            
            // Show notification
            if (GameManager.Instance?.uiManager != null)
            {
                GameManager.Instance.uiManager.ShowKingdomNotification(kingdom.name, "Kingdom Restored!");
            }
        }
        else if (kingdom.restorationProgress > 0)
        {
            kingdom.status = KingdomStatus.RESTORING;
            Debug.Log($"🔨 {kingdom.name} restoration: {kingdom.restorationProgress:F1}%");
        }
    }
    
    public KingdomStats GetStats()
    {
        return new KingdomStats
        {
            total = kingdoms.Count,
            discovered = kingdoms.Count(k => k.status != KingdomStatus.FALLEN),
            vaultsCleared = kingdoms.Count(k => k.vaultCleared),
            restored = kingdoms.Count(k => k.status == KingdomStatus.RESTORED),
            progress = kingdoms.Average(k => k.restorationProgress)
        };
    }
    
    public bool IsCrownConvergenceUnlocked()
    {
        // Crown Convergence unlocks after clearing 7 other kingdoms
        return kingdoms.Count(k => k.vaultCleared && k.id != "crown_convergence") >= 7;
    }
}

[System.Serializable]
public class Kingdom
{
    public string id;
    public string name;
    public KingdomElement element;
    public KingdomStatus status;
    public float restorationProgress;
    public bool vaultCleared;
    public bool cutsceneWatched;
    public string lore;
    
    public Color GetColor()
    {
        switch (element)
        {
            case KingdomElement.NATURE: return new Color(0.18f, 0.8f, 0.44f);
            case KingdomElement.FIRE: return new Color(0.91f, 0.3f, 0.24f);
            case KingdomElement.ICE: return new Color(0.2f, 0.6f, 0.86f);
            case KingdomElement.DESERT: return new Color(0.95f, 0.61f, 0.07f);
            case KingdomElement.POISON: return new Color(0.61f, 0.35f, 0.71f);
            case KingdomElement.LIGHT: return new Color(0.93f, 0.94f, 0.95f);
            case KingdomElement.SHADOW: return new Color(0.2f, 0.29f, 0.37f);
            case KingdomElement.UNIFIED: return Color.white;
            default: return Color.gray;
        }
    }
}

public enum KingdomElement
{
    NATURE,
    FIRE,
    ICE,
    DESERT,
    POISON,
    LIGHT,
    SHADOW,
    UNIFIED
}

public enum KingdomStatus
{
    FALLEN,
    DISCOVERED,
    VAULT_CLEARED,
    RESTORING,
    RESTORED
}

public struct KingdomStats
{
    public int total;
    public int discovered;
    public int vaultsCleared;
    public int restored;
    public float progress;
}

File: prototypes/src/systems/KingdomManager.ts
/**
 * Kingdom Manager - Manages all 8 kingdoms and their restoration
 */

import { Kingdom, KingdomElement } from '../core/Kingdom';
import { Vault, VaultManager } from './VaultSystem';

export class KingdomManager {
  private kingdoms: Map<string, Kingdom> = new Map();
  private vaultManager: VaultManager;
  private totalRestorationProgress: number = 0;

  constructor() {
    this.vaultManager = new VaultManager();
    this.initializeKingdoms();
  }

  /**
   * Initialize all 8 kingdoms
   */
  private initializeKingdoms(): void {
    const kingdomsData = [
      {
        id: 'sylvara',
        name: 'Sylvara',
        element: KingdomElement.NATURE,
        lore: {
          description: 'A thriving woodland empire where druids lived in harmony with nature',
          downfall: 'Corrupted by Shadow cultists who poisoned the Great Rootspire',
          vaultLocation: 'Beneath the Rootspire ancient tree palace'
        }
      },
      {
        id: 'pyrrathia',
        name: 'Pyrrathia',
        element: KingdomElement.FIRE,
        lore: {
          description: 'A volcanic empire built inside an active supervolcano',
          downfall: 'Destroyed by catastrophic Water flood from neighboring kingdom',
          vaultLocation: 'Deep within the dormant volcano magma chamber'
        }
      },
      {
        id: 'frostvale',
        name: 'Frostvale',
        element: KingdomElement.ICE,
        lore: {
          description: 'A crystalline ice kingdom built on eternal glaciers',
          downfall: 'Frozen in time by Eternal Winter spell during dragon attack',
          vaultLocation: 'Top of the frozen palace, inside queen throne room'
        }
      },
      {
        id: 'zerath_dunes',
        name: 'Zerath Dunes',
        element: KingdomElement.DESERT,
        lore: {
          description: 'A golden desert empire powered by sun magic',
          downfall: 'Buried by massive sandstorms after water drought',
          vaultLocation: 'Deep beneath the buried Pyramid of the Sun King'
        }
      },
      {
        id: 'morgahl_fen',
        name: "Mor'gahl Fen",
        element: KingdomElement.POISON,
        lore: {
          description: 'A mysterious swamp civilization that mastered poison alchemy',
          downfall: 'Poison disaster followed by Light Kingdom purge',
          vaultLocation: 'Sunken Palace beneath the toxic swamp'
        }
      },
      {
        id: 'luminaris',
        name: 'Luminaris',
        element: KingdomElement.LIGHT,
        lore: {
          description: 'A radiant kingdom built from enchanted crystals',
          downfall: 'Shattered by massive light explosion during Shadow invasion',
          vaultLocation: 'Highest fragment of the floating Shardspire'
        }
      },
      {
        id: 'umbral_reach',
        name: 'Umbral Reach',
        element: KingdomElement.SHADOW,
        lore: {
          description: 'A dark kingdom built in eternal twilight',
          downfall: 'Pulled into void dimension during Light Kingdom holy war',
          vaultLocation: 'Deep in underground abyss, phasing between dimensions'
        }
      },
      {
        id: 'crown_convergence',
        name: 'Crown Convergence',
        element: KingdomElement.UNIFIED,
        lore: {
          description: 'The original kingdom where the Crown was forged',
          downfall: 'Torn apart when all seven kingdoms attacked simultaneously',
          vaultLocation: 'Scattered across floating sky islands'
        }
      }
    ];

    kingdomsData.forEach(data => {
      const kingdom = new Kingdom(data.id, data.name, data.element, data.lore);
      this.kingdoms.set(data.id, kingdom);
      
      // Create vault for each kingdom
      this.vaultManager.createVault(kingdom);
    });

    console.log('🏰 Initialized 8 kingdoms');
  }

  /**
   * Get kingdom by ID
   */
  public getKingdom(kingdomId: string): Kingdom | undefined {
    return this.kingdoms.get(kingdomId);
  }

  /**
   * Get all kingdoms
   */
  public getAllKingdoms(): Kingdom[] {
    return Array.from(this.kingdoms.values());
  }

  /**
   * Get kingdoms by status
   */
  public getKingdomsByStatus(status: string): Kingdom[] {
    return this.getAllKingdoms().filter(k => k.status === status);
  }

  /**
   * Get vault for kingdom
   */
  public getVault(kingdomId: string): Vault | undefined {
    return this.vaultManager.getVault(kingdomId);
  }

  /**
   * Discover kingdom
   */
  public discoverKingdom(kingdomId: string): void {
    const kingdom = this.getKingdom(kingdomId);
    if (kingdom) {
      kingdom.discover();
      this.updateTotalProgress();
    }
  }

  /**
   * Clear kingdom vault
   */
  public clearVault(kingdomId: string): void {
    const kingdom = this.getKingdom(kingdomId);
    const vault = this.getVault(kingdomId);
    
    if (kingdom && vault) {
      vault.open();
      const loot = vault.collectTreasure();
      kingdom.clearVault();
      this.updateTotalProgress();
      
      console.log(`👑 ${kingdom.name} vault cleared! Collected ${loot.length} items`);
    }
  }

  /**
   * Restore kingdom
   */
  public restoreKingdom(kingdomId: string, amount: number): void {
    const kingdom = this.getKingdom(kingdomId);
    if (kingdom) {
      kingdom.restore(amount);
      this.updateTotalProgress();
    }
  }

  /**
   * Update total restoration progress
   */
  private updateTotalProgress(): void {
    const kingdoms = this.getAllKingdoms();
    const totalProgress = kingdoms.reduce((sum, k) => sum + k.restorationProgress, 0);
    this.totalRestorationProgress = totalProgress / kingdoms.length;
    
    console.log(`🌍 Total restoration progress: ${this.totalRestorationProgress.toFixed(1)}%`);
  }

  /**
   * Get total restoration progress
   */
  public getTotalProgress(): number {
    return this.totalRestorationProgress;
  }

  /**
   * Get discovered kingdoms count
   */
  public getDiscoveredCount(): number {
    return this.getAllKingdoms().filter(k => k.isAccessible()).length;
  }

  /**
   * Get cleared vaults count
   */
  public getClearedVaultsCount(): number {
    return this.vaultManager.getClearedCount();
  }

  /**
   * Get restored kingdoms count
   */
  public getRestoredCount(): number {
    return this.getKingdomsByStatus('restored').length;
  }

  /**
   * Check if all kingdoms are restored
   */
  public isAllRestored(): boolean {
    return this.getRestoredCount() === 8;
  }

  /**
   * Get next kingdom to discover (based on difficulty)
   */
  public getNextKingdom(): Kingdom | undefined {
    const order = [
      'sylvara',
      'pyrrathia',
      'frostvale',
      'zerath_dunes',
      'morgahl_fen',
      'luminaris',
      'umbral_reach',
      'crown_convergence'
    ];

    for (const id of order) {
      const kingdom = this.getKingdom(id);
      if (kingdom && !kingdom.isAccessible()) {
        return kingdom;
      }
    }

    return undefined;
  }

  /**
   * Get kingdom statistics
   */
  public getStats(): {
    total: number;
    discovered: number;
    vaultsCleared: number;
    restored: number;
    progress: number;
  } {
    return {
      total: 8,
      discovered: this.getDiscoveredCount(),
      vaultsCleared: this.getClearedVaultsCount(),
      restored: this.getRestoredCount(),
      progress: this.totalRestorationProgress
    };
  }

  /**
   * Play kingdom cutscene
   */
  public playCutscene(kingdomId: string): void {
    const kingdom = this.getKingdom(kingdomId);
    if (kingdom && !kingdom.cutsceneWatched) {
      console.log(`🎬 Playing ${kingdom.name} cutscene...`);
      console.log(`📖 ${kingdom.lore.description}`);
      console.log(`💔 Downfall: ${kingdom.lore.downfall}`);
      console.log(`🏛️ Vault: ${kingdom.lore.vaultLocation}`);
      
      kingdom.watchCutscene();
      
      // Bonus for watching cutscene
      console.log('🎁 Cutscene bonus: +500 XP, +100g');
    }
  }

  /**
   * Get kingdom by element
   */
  public getKingdomByElement(element: KingdomElement): Kingdom | undefined {
    return this.getAllKingdoms().find(k => k.element === element);
  }

  /**
   * Check if Crown Convergence is unlocked
   */
  public isCrownConvergenceUnlocked(): boolean {
    // Crown Convergence unlocks after clearing all 7 other kingdoms
    return this.getClearedVaultsCount() >= 7;
  }
}

// Singleton instance
export const kingdomManager = new KingdomManager();
